TensorFlow master build fails on s390x due to boringssl issue

<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

Reference issues: https://github.com/tensorflow/tensorflow/issues/14039, https://github.com/tensorflow/tensorflow/issues/50351

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): source 
- TensorFlow version: 2.7.0 (from master branch)
- Python version: 3.6.9
- Installed using virtualenv? pip? conda?: No
- Bazel version (if compiling from source): bazel 3.7.2- (@non-git)
- GCC/Compiler version (if compiling from source): gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the problem**

TensorFlow master build is failing on `s390x` due to `boringssl` issue. It looks like some recent commits in master branch have triggered it. 

I made some tests and it looks like this commit (https://github.com/tensorflow/tensorflow/commit/738e5aa37935f713a2366fc00a26d1d5830cb971) trigged the issue (the previous commit didn't trigger it). This commit has the dependency of `@com_github_grpc_grpc` which needs `boringssl` and the code appears to be in the `experimental` folder.

Since `boringssl` is still not supported on `s390x`, is there any way to disable this part on `s390x` build? Thank you very much!

**Provide the exact sequence of commands / steps that you executed before running into the problem**

`bazel --host_jvm_args="-Xms1024m" --host_jvm_args="-Xmx2048m" build  --define=tensorflow_mkldnn_contraction_kernel=0 //tensorflow/tools/pip_package:build_pip_package`

or

`bazel --host_jvm_args="-Xms1024m" --host_jvm_args="-Xmx2048m" build  --define=tensorflow_mkldnn_contraction_kernel=0 --config=noaws //tensorflow/tools/pip_package:build_pip_package`

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.
```bash
ERROR: /home/test/Tensorflow_tmp/_bazel_test/78e38be8987e210d400487db993cf28d/external/com_github_grpc_grpc/BUILD:2013:16: C++ compilation of rule '@com_github_grpc_grpc//:tsi' failed (Exit 1): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 46 argument(s) skipped)
In file included from external/boringssl/src/include/openssl/ssl.h:145:0,
                 from external/com_github_grpc_grpc/src/core/tsi/ssl/session_cache/ssl_session.h:29,
                 from external/com_github_grpc_grpc/src/core/tsi/ssl/session_cache/ssl_session_cache.cc:23:
external/boringssl/src/include/openssl/base.h:122:2: error: #error "Unknown target CPU"
 #error "Unknown target CPU"
  ^~~~~
In file included from external/boringssl/src/include/openssl/asn1.h:68:0,
                 from external/boringssl/src/include/openssl/x509.h:70,
                 from external/boringssl/src/include/openssl/pem.h:67,
                 from external/boringssl/src/include/openssl/ssl.h:149,
                 from external/com_github_grpc_grpc/src/core/tsi/ssl/session_cache/ssl_session.h:29,
                 from external/com_github_grpc_grpc/src/core/tsi/ssl/session_cache/ssl_session_cache.cc:23:
external/boringssl/src/include/openssl/bn.h:165:2: error: #error "Must define either OPENSSL_32_BIT or OPENSSL_64_BIT"
 #error "Must define either OPENSSL_32_BIT or OPENSSL_64_BIT"
```
```bash
ERROR: /home/test/Tensorflow_tmp/_bazel_test/78e38be8987e210d400487db993cf28d/external/boringssl/BUILD:130:11: C++ compilation of rule '@boringssl//:crypto' failed (Exit 1): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 24 argument(s) skipped)
In file included from external/boringssl/src/include/openssl/digest.h:60:0,
                 from external/boringssl/src/crypto/x509v3/v3_skey.c:61:
external/boringssl/src/include/openssl/base.h:122:2: error: #error "Unknown target CPU"
 #error "Unknown target CPU"
  ^~~~~
In file included from external/boringssl/src/include/openssl/asn1.h:68:0,
                 from external/boringssl/src/include/openssl/x509.h:70,
                 from external/boringssl/src/include/openssl/x509v3.h:60,
                 from external/boringssl/src/crypto/x509v3/v3_skey.c:64:
external/boringssl/src/include/openssl/bn.h:165:2: error: #error "Must define either OPENSSL_32_BIT or OPENSSL_64_BIT"
 #error "Must define either OPENSSL_32_BIT or OPENSSL_64_BIT"
```

--- a/configure.py
+++ b/configure.py
@@ -83,6 +83,8 @@ def is_macos():
 def is_ppc64le():
   return platform.machine() == 'ppc64le'
 
+def is_s390x():
+  return platform.machine() == 's390x'
 
 def is_cygwin():
   return platform.system().startswith('CYGWIN_NT')
@@ -1007,6 +1009,10 @@ def system_specific_test_config(environ_cp):
 
 def set_system_libs_flag(environ_cp):
   syslibs = environ_cp.get('TF_SYSTEM_LIBS', '')
+  
+  if is_s390x() and "boringssl" not in syslibs:
+    syslibs = "boringssl" + (", " + syslibs if syslibs != "" else "")
+
   if syslibs:
     if ',' in syslibs:
       syslibs = ','.join(sorted(syslibs.split(',')))

--- a/tensorflow/BUILD
+++ b/tensorflow/BUILD
@@ -1086,7 +1086,6 @@ cc_library(
     name = "grpc",
     visibility = ["//visibility:public"],
     deps = select({
-        ":linux_s390x": ["@com_github_grpc_grpc//:grpc_unsecure"],
         "//conditions:default": ["@com_github_grpc_grpc//:grpc"],
     }),
 )
@@ -1095,7 +1094,6 @@ cc_library(
     name = "grpc++",
     visibility = ["//visibility:public"],
     deps = select({
-        ":linux_s390x": ["@com_github_grpc_grpc//:grpc++_unsecure"],
         "//conditions:default": ["@com_github_grpc_grpc//:grpc++"],
     }),
 )

--- a/tensorflow/tsl/BUILD
+++ b/tensorflow/tsl/BUILD
@@ -467,7 +467,6 @@ cc_library(
     name = "grpc++",
     visibility = ["//visibility:public"],
     deps = select({
-        ":linux_s390x": ["@com_github_grpc_grpc//:grpc++_unsecure"],
         "//conditions:default": ["@com_github_grpc_grpc//:grpc++"],
     }),
 )

--- a/tensorflow/tsl/platform/default/build_config.bzl
+++ b/tensorflow/tsl/platform/default/build_config.bzl
@@ -243,7 +243,6 @@ def cc_proto_library(
 
     if use_grpc_plugin:
         cc_libs += select({
-            clean_dep("//tensorflow/tsl:linux_s390x"): ["//external:grpc_lib_unsecure"],
             "//conditions:default": ["//external:grpc_lib"],
         })
 
@@ -326,7 +325,6 @@ def cc_grpc_library(
     proto_targets += srcs
 
     extra_deps += select({
-        clean_dep("//tensorflow/tsl:linux_s390x"): ["//external:grpc_lib_unsecure"],
         "//conditions:default": ["//external:grpc_lib"],
     })
 

